subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java

```diff
@@ -391,42 +391,34 @@ public FileCollection call() throws Exception {
     }
 
     private void setupScopes(GenerateIdeaModule ideaModule) {
-        Map<String, Map<String, Collection<Configuration>>> scopes = Maps.newLinkedHashMap();
-        for (GeneratedIdeaScope scope : GeneratedIdeaScope.values()) {
-            Map<String, Collection<Configuration>> plusMinus = Maps.newLinkedHashMap();
-            plusMinus.put(IdeaDependenciesProvider.SCOPE_PLUS, Lists.<Configuration>newArrayList());
-            plusMinus.put(IdeaDependenciesProvider.SCOPE_MINUS, Lists.<Configuration>newArrayList());
-            scopes.put(scope.name(), plusMinus);
-        }
 
         Project project = ideaModule.getProject();
-        ConfigurationContainer configurations = project.getConfigurations();
+        IdeaScopesConfig scopes = new IdeaScopesConfig(project.getConfigurations());
+
+        scopes.named(GeneratedIdeaScope.PROVIDED)
+            .plus(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME);
 
-        Collection<Configuration> provided = scopes.get(GeneratedIdeaScope.PROVIDED.name()).get(IdeaDependenciesProvider.SCOPE_PLUS);
-        provided.add(configurations.getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME));
+        scopes.named(GeneratedIdeaScope.RUNTIME)
+            .plus(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME);
 
-        Collection<Configuration> runtime = scopes.get(GeneratedIdeaScope.RUNTIME.name()).get(IdeaDependenciesProvider.SCOPE_PLUS);
-        runtime.add(configurations.getByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME));
+        scopes.named(GeneratedIdeaScope.TEST)
+            .plus(JavaPlugin.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME)
+            .plus(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME)
+            .minus(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME);
```

Comment:
Does TEST in idea also extend RUNTIME? If so, you could also do `minus(RUNTIME_CLASSPATH)` here.

Reply:
Idea's scope do not extend each other, but rather contribute to a classpath. To be exact, RUNTIME scope deps are added to _tests execution_ classpath.  `minus(RUNTIME_CLASSPATH)` will remove duplicate deps coming from `runtimeOnly` but that's not a big number usually.

Reply:
It's still something, I'd add it :)